r/tests/testthat/test-dplyr-funcs-datetime.R

```diff
@@ -616,3 +616,292 @@ test_that("extract yday from date", {
+        out_1 = round_date(datetime, ".01 second"),
+        out_2 = round_date(datetime, ".001 second"),
+        out_3 = round_date(datetime, ".00001 second")
+      ) %>%
+      collect(),
+    test_df
+  )
+})
+
+
+# a simplified test case using UTC timezone only
+test_df_v2 <- tibble::tibble(
+  datetime = c(as.POSIXct("2017-01-01 00:00:11.3456789", tz = "UTC"), NA),
+  date = c(as.Date("2021-09-09"), NA),
+  integer = 1:2
+)
+
+test_that("datetime round/floor/ceil to month/quarter/year", {
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = round_date(datetime, "month"),
+        out_2 = round_date(datetime, "quarter"),
+        out_3 = round_date(datetime, "year"),
+      ) %>%
+      collect(),
+    test_df_v2
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = floor_date(datetime, "month"),
+        out_2 = floor_date(datetime, "quarter"),
+        out_3 = floor_date(datetime, "year"),
+      ) %>%
+      collect(),
+    test_df_v2
+  )
+
+  compare_dplyr_binding(
+    .input %>%
+      mutate(
+        out_1 = ceiling_date(datetime, "month"),
+        out_2 = ceiling_date(datetime, "quarter"),
+        out_3 = ceiling_date(datetime, "year"),
+      ) %>%
+      collect(),
+    test_df_v2
+  )
+})
+
+# NOTE: the hard coding of week_starts = 4 needs to be fixed. round_temporal()
+# treats 1970-01-01 as the beginning of week 1, i.e., week_starts on a Thursday
+
+# NOTE: arrow dplyr binding for ceiling_date() does not force dates up to the
+# next date. the logic mirrors lubridate prior to v1.6.0 (change_on_boundary = FALSE).
+# I'm not 100% sold on this implementation, but it's not obviously terrible
+
+test_that("datetime round/floor/ceil to week", {
+
+  expect_equal(
+    test_df_v2 %>%
+      arrow_table() %>%
+      mutate(out = round_date(datetime, "1 week")) %>%
+      collect(),
+    test_df_v2 %>%
+      mutate(out = round_date(datetime, "1 week", week_start = 4))
+  )
+
+  expect_equal(
+    test_df_v2 %>%
+      arrow_table() %>%
+      mutate(out = ceiling_date(datetime, "1 week")) %>%
+      collect(),
+    test_df_v2 %>%
+      mutate(out = ceiling_date(datetime, "1 week", week_start = 4, change_on_boundary = FALSE))
+  )
+
+  expect_equal(
+    test_df_v2 %>%
+      arrow_table() %>%
+      mutate(out = floor_date(datetime, "1 week")) %>%
+      collect(),
+    test_df_v2 %>%
+      mutate(out = floor_date(datetime, "1 week", week_start = 4))
+  )
+})
+
+# NOTE: lubridate::round_date() sometimes coerces output from Date to POSIXct.
+# this is not the default for the round_temporal() function in libarrow, which
+# is type stable: timestamps stay timestamps, and date32 stays date32. the
+# current implementation preserves the type stability property. consequently
+# there are edge cases where the arrow dplyr binding will not precisely mirror
+# the lubridate original. with that in mind, all tests for date32 rounding coerce
+# the lubridate equivalent back to Date
```

Comment:
Yup, I agree here. It might be nice to add a small comment in those test cases about this (so future-us remembers why we have `as.Date()` at the end there